lib/thesis/controller.ex

```diff
@@ -23,6 +23,53 @@ defmodule Thesis.Controller do
   defmacro __using__(_) do
     quote do
       plug Thesis.Controller.Plug
+
+      @doc """
+      Renders a page that exists only in the database, as opposed to a page that
+      is defined in the router.
+
+      This function will also redirect to the page's `redirect_url` if it is set.
+      However, if the user is in edit mode, we will render the default template and
+      display an alert telling them about the redirect. This gives the user a chance
+      to edit where this page redirects to.
+      """
+      def render_dynamic(conn, opts \\ []) do
+        page = conn.assigns[:thesis_page]
+        if page && page.id do
+          do_render_dynamic(conn, page, opts)
+        else
+          render_not_found(conn)
+        end
+      end
```

Comment:
Here's the `render_dynamic(conn)` function I mentioned above. This part of the function checks if there's a Thesis page and that it has an id (it is already stored in the database). If not, render a 404 page.